pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorService.java

```diff
@@ -121,7 +122,11 @@ private Comparator<Object[]> getTypeCompatibleComparator(List<OrderByExpressionC
         Object v2 = o2[index];
         int result;
         if (isNumber[i]) {
-          result = Double.compare(((Number) v1).doubleValue(), ((Number) v2).doubleValue());
+          if (columnDataTypes[i] == ColumnDataType.BIG_DECIMAL) {
```

Comment:
We may avoid this extra if check per comparison by excluding `BIG_DECIMAL` from double comparison (change `isNumber` to `useDoubleComparison`)